README
This is a reverse engineered libomap3camd interface
Using this, it should be possible to rebuild subdevsrc for nemo mobile.


debian/changelog
libomap3camd (0.0.0) unstable; urgency=low

  * Initial release.

 -- Mohammed Sameer <msameer@foolab.org>  Mon, 08 Apr 2013 23:33:41 +0300


app.c
/*!
 *
 * Copyright (C) 2013 Mohammed Sameer <msameer@foolab.org>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <gst/gst.h>
#include <glib.h>
#include <dlfcn.h>
#include <stdarg.h>
#define GST_USE_UNSTABLE_API
#include <gst/interfaces/photography.h>

// Launch app under gdb.
// Wait for OK then set a breakpoint on cam_feature_set
// info registers:
// r0 -> libomap3camd handle
// r1 -> first argument passed to cam_feature_set
// r2 -> get that and match from subdevsrc maps.

//  gcc -o app app.c `pkg-config gstreamer-0.10 gstreamer-plugins-bad-0.10 --cflags --libs`

typedef struct {
  const gchar *id;
  const char *prop;
  const char *name;
  int max_val;
  gboolean (*func)(void *test_data);
} test;

typedef struct {
  GstElement *src;
  test *t;
  GMainLoop *loop;
  GstPhotography *photo;
} test_data;

gboolean video_scene(test_data *data) {
  g_object_set(data->src, "viewfinder-mode", 1, NULL);

  return do_test(data);
}

gboolean video_torch(test_data *data) {
  puts("Video torch");
  g_object_set(data->src, "viewfinder-mode", 1, NULL);

  puts("Enabling");
  g_object_set(data->src, "video-torch", TRUE, NULL);

  puts("Disabling");
  g_object_set(data->src, "video-torch", FALSE, NULL);

  g_main_loop_quit(data->loop);

  return FALSE; // bye!
}

gboolean capture(test_data *data) {
  puts("Capture");

  GstCaps *mode_caps = gst_caps_new_simple ("video/x-raw-yuv",
				   "format", GST_TYPE_FOURCC, GST_MAKE_FOURCC ('U', 'Y', 'V', 'Y'),
				   "width", G_TYPE_INT, 640,
				   "height", G_TYPE_INT, 480,
				   "framerate", GST_TYPE_FRACTION, 2406, 100,
				   NULL);

  GstOperationMode op_mode = GST_PHOTOGRAPHY_OPERATION_MODE_IMAGE_CAPTURE;
  gst_photography_set_format(data->photo, op_mode, mode_caps);

  g_main_loop_quit(data->loop);

  return FALSE; // bye!
}

gboolean autofocus(test_data *data) {
  puts("autofocus");

  puts("Enabling");
  gst_photography_set_autofocus(data->photo, TRUE);

  puts("Disabling");
  gst_photography_set_autofocus(data->photo, FALSE);

  g_main_loop_quit(data->loop);

  return FALSE; // bye!
}

test tests[] = {
  {"wb", "white-balance-mode", "White balance", 5, NULL},
  {"flash", "flash-mode", "Flash", 4, NULL},
  {"tone", "colour-tone-mode", "color tone", 11, NULL},
  {"flicker", "flicker-mode", "flicker mode", 3, NULL},
  {"focus", "focus-mode", "Focus mode", 8, NULL},
  {"scene", "scene-mode", "Image scene", 6, NULL},
  {"vs", "scene-mode", "Video scene", 6, video_scene},
  {"torch", "", "", -1, video_torch},
  {"zoom", "zoom", "Zoom", 3, NULL},
  {"capture", "", "", -1, capture},
  {"af", "", "", -1, autofocus},

  {NULL, NULL, NULL, 0},
};

gboolean do_test(test_data *data) {
  int x = 0;

  puts(data->t->name);

  for (x = 0; x < data->t->max_val + 1; x++) {
    printf("Setting %s to %d\n", data->t->prop, x);
    g_object_set(data->src, data->t->prop, x, NULL);
  }

  g_main_loop_quit(data->loop);

  return FALSE; // bye!
}

int get_test(int argc, char *argv[]) {
  int x = 0;
  if (argc != 2) {
    return -1;
  }

  int len = sizeof(tests) / sizeof(test);
  for (x = 0; x < len; x++) {
    if (!strcmp(tests[x].id, argv[1])) {
      return x;
    }
  }

  return -1;
}

int main(int argc, char *argv[]) {
  int index = get_test(argc, argv);
  if (index == -1) {
    printf("Test required\n");
    return 1;
  }

  gst_init(&argc, &argv);

  GMainLoop *loop = g_main_loop_new(NULL, FALSE);

  GstElement *pipe = gst_pipeline_new(NULL);
  GstElement *src = gst_element_factory_make("subdevsrc", NULL);
  GstElement *sink = gst_element_factory_make("fakesink", NULL);

  gst_bin_add_many(GST_BIN(pipe), src, sink, NULL);
  gst_element_link(src, sink);

  gst_element_set_state(pipe, GST_STATE_PLAYING);

  puts("Set break point now!");

  test *t = &tests[index];

  test_data data;
  data.src = src;
  data.t = t;
  data.loop = loop;
  data.photo = GST_PHOTOGRAPHY(src);

  if (t->func) {
    g_timeout_add(2000, (GSourceFunc)t->func, &data);
  }
  else {
    g_timeout_add(2000, (GSourceFunc)do_test, &data);
  }

  g_main_loop_run(loop);

  gst_element_set_state(pipe, GST_STATE_NULL);

  return 0;
}


omap3cam.c
/*!
 *
 * Copyright (C) 2013 Mohammed Sameer <msameer@foolab.org>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include "omap3cam.h"
#include <stdlib.h>
#include <stdio.h>

void *cam_library_create() {
  fprintf(stderr, "This library should not be used\n");
  abort();

  return NULL;
}

void cam_library_delete(void *handle) {
  fprintf(stderr, "This library should not be used\n");
  abort();
}

cam_err_t cam_feature_set(void *handle, cam_feature_t feature, ...) {
  fprintf(stderr, "This library should not be used\n");
  abort();

  return CAM_ERR_INTERNAL;
}

cam_err_t cam_feature_get(void *handle, cam_feature_t feature, ...) {
  fprintf(stderr, "This library should not be used\n");
  abort();

  return CAM_ERR_INTERNAL;
}


linux/media.h
/*
 * Multimedia device API
 *
 * Copyright (C) 2010 Nokia Corporation
 *
 * Contacts: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
 *	     Sakari Ailus <sakari.ailus@maxwell.research.nokia.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifndef __LINUX_MEDIA_H
#define __LINUX_MEDIA_H

#include <linux/ioctl.h>
#include <linux/types.h>
#include <linux/version.h>

#define MEDIA_API_VERSION	KERNEL_VERSION(0, 1, 0)

struct media_device_info {
	char driver[16];
	char model[32];
	char serial[40];
	char bus_info[32];
	__u32 media_version;
	__u32 hw_revision;
	__u32 driver_version;
	__u32 reserved[31];
};

#define MEDIA_ENT_ID_FLAG_NEXT			(1 << 31)

#define MEDIA_ENT_TYPE_SHIFT			16
#define MEDIA_ENT_TYPE_MASK			0x00ff0000
#define MEDIA_ENT_SUBTYPE_MASK			0x0000ffff

#define MEDIA_ENT_T_DEVNODE			(1 << MEDIA_ENT_TYPE_SHIFT)
#define MEDIA_ENT_T_DEVNODE_V4L			(MEDIA_ENT_T_DEVNODE + 1)
#define MEDIA_ENT_T_DEVNODE_FB			(MEDIA_ENT_T_DEVNODE + 2)
#define MEDIA_ENT_T_DEVNODE_ALSA		(MEDIA_ENT_T_DEVNODE + 3)
#define MEDIA_ENT_T_DEVNODE_DVB			(MEDIA_ENT_T_DEVNODE + 4)

#define MEDIA_ENT_T_V4L2_SUBDEV			(2 << MEDIA_ENT_TYPE_SHIFT)
#define MEDIA_ENT_T_V4L2_SUBDEV_SENSOR		(MEDIA_ENT_T_V4L2_SUBDEV + 1)
#define MEDIA_ENT_T_V4L2_SUBDEV_FLASH		(MEDIA_ENT_T_V4L2_SUBDEV + 2)
#define MEDIA_ENT_T_V4L2_SUBDEV_LENS		(MEDIA_ENT_T_V4L2_SUBDEV + 3)

#define MEDIA_ENT_FL_DEFAULT			(1 << 0)

struct media_entity_desc {
	__u32 id;
	char name[32];
	__u32 type;
	__u32 revision;
	__u32 flags;
	__u32 group_id;
	__u16 pads;
	__u16 links;

	__u32 reserved[4];

	union {
		/* Node specifications */
		struct {
			__u32 major;
			__u32 minor;
		} v4l;
		struct {
			__u32 major;
			__u32 minor;
		} fb;
		struct {
			__u32 card;
			__u32 device;
			__u32 subdevice;
		} alsa;
		int dvb;

		/* Sub-device specifications */
		/* Nothing needed yet */
		__u8 raw[184];
	};
};

#define MEDIA_PAD_FL_SINK			(1 << 0)
#define MEDIA_PAD_FL_SOURCE			(1 << 1)

struct media_pad_desc {
	__u32 entity;		/* entity ID */
	__u16 index;		/* pad index */
	__u32 flags;		/* pad flags */
	__u32 reserved[2];
};

#define MEDIA_LNK_FL_ENABLED			(1 << 0)
#define MEDIA_LNK_FL_IMMUTABLE			(1 << 1)
#define MEDIA_LNK_FL_DYNAMIC			(1 << 2)

struct media_link_desc {
	struct media_pad_desc source;
	struct media_pad_desc sink;
	__u32 flags;
	__u32 reserved[2];
};

struct media_links_enum {
	__u32 entity;
	/* Should have enough room for pads elements */
	struct media_pad_desc *pads;
	/* Should have enough room for links elements */
	struct media_link_desc *links;
	__u32 reserved[4];
};

#define MEDIA_IOC_DEVICE_INFO           _IOWR('|', 0x00, struct media_device_info)
#define MEDIA_IOC_ENUM_ENTITIES         _IOWR('|', 0x01, struct media_entity_desc)
#define MEDIA_IOC_ENUM_LINKS            _IOWR('|', 0x02, struct media_links_enum)
#define MEDIA_IOC_SETUP_LINK            _IOWR('|', 0x03, struct media_link_desc)

/* Compatibility stuff */
#define MEDIA_ENTITY_TYPE_NODE_V4L      MEDIA_ENT_T_DEVNODE_V4L
#define MEDIA_ENTITY_TYPE_SUBDEV        MEDIA_ENT_T_V4L2_SUBDEV
#define MEDIA_ENTITY_TYPE_MASK          MEDIA_ENT_TYPE_MASK
#define MEDIA_ENTITY_TYPE_NODE          MEDIA_ENT_T_DEVNODE
#define MEDIA_LINK_FLAG_ACTIVE          MEDIA_LNK_FL_ENABLED
#define MEDIA_LINK_FLAG_IMMUTABLE       MEDIA_LNK_FL_IMMUTABLE
#define MEDIA_PAD_FLAG_OUTPUT           MEDIA_PAD_FL_SOURCE
#define MEDIA_PAD_FLAG_INPUT            MEDIA_PAD_FL_SINK

#endif /* __LINUX_MEDIA_H */


omap3cam.h
/*!
 *
 * Copyright (C) 2013 Mohammed Sameer <msameer@foolab.org>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#ifndef __OMAP3_CAM_H__
#define __OMAP3_CAM_H__


#define CAM_IMAGE_INFO             39
#define CAM_VIEWFINDER_MODE_VIDEO_RECORD 1
#define CAM_VIEWFINDER_MODE_STILL_CAPTURE 0
#define CAM_MOTION_DETECTION 33
#define CAM_ENGINE_CONTROL 35
#define CAM_MAKER_NOTE 38
#define CAM_IPP_PARAMETERS 41
#define CAM_EXPOSURE_SHAKE_RISK 14
#define CAM_LIGHT_FLASH_STATUS 43
#define CAM_FOCUS_STATUS 23
#define CAM_EXPOSURE_MODE_AUTO 0
#define CAM_LIGHT_FLASH_REQUIRED 44
#define OBJECT_TRACKING_MAX_COUNT 16

typedef enum {
  CAM_ENGINE_VIEWFINDER = 0,
  CAM_ENGINE_STILL_CAPTURE = 1,
  CAM_ENGINE_SNAP_SHOT = 2,
} cam_engine_t;

typedef enum {
  CAM_EXPOSURE_ISO_VALUE_800 = 4,
  CAM_EXPOSURE_ISO_VALUE_400 = 3,
  CAM_EXPOSURE_ISO_VALUE_200 = 2,
  CAM_EXPOSURE_ISO_VALUE_100 = 1,
  CAM_EXPOSURE_ISO_VALUE_AUTO = 0,
} cam_exposure_iso_value_t;

typedef enum {
  CAM_FRAME_FORMAT_YUYV = 0,
  CAM_FRAME_FORMAT_UYVY = 1,
  CAM_FRAME_FORMAT_BAYER10 = 2,
  CAM_PIXEL_FORMAT_YUYV = 0,
  CAM_PIXEL_FORMAT_UYVY = 1,
  CAM_PIXEL_FORMAT_BAYER10 = 2,
} cam_pixel_format_t;

typedef enum {
  CAM_ERR_NONE = 0,
  CAM_ERR_PARAM = 1,
  CAM_ERR_UNSUPP = 2,
  CAM_ERR_HW = 3,
  CAM_ERR_SYS = 4,
  CAM_ERR_LEXIT = 5,
  CAM_ERR_DEPRECATED = 6,
  CAM_ERR_INVALID_STATE = 7,
  CAM_ERR_INTERNAL = 8
} cam_err_t;

typedef enum {
  CAM_STILL_CAPTURE_START = 6,
  CAM_STILL_CAPTURE_DONE = 7,
  CAM_ALGO_WHITE_BALANCE = 0,
  CAM_ALGO_EXPOSURE = 1,
  CAM_ALGO_FOCUS_SINGLE = 2,
  CAM_ALGO_FOCUS_CONTINUOUS = 5,
  CAM_ALGO_SHAKE_RISK = 9,
  CAM_ALGO_LINE_FREQ_DETECTED = 8,
  CAM_ALGO_FLASHLIGHT_CONTROL = 10,
  CAM_ALGO_FLASHLIGHT_REQUIRED = 11,
} cam_algo_id_t;

typedef enum {
  CAM_CAMERA_DEVICE = 1,
  CAM_EXPOSURE_MODE = 12,
  CAM_AWB_MODE = 25,
  CAM_FOCUS_MODE = 20,
  CAM_LIGHT_PRIVACY_MODE = 11,
  CAM_LIGHT_ASSIST_MODE = 10,
  CAM_GENERAL_FLICKER_REDUCTION_MODE = 8,
  CAM_GENERAL_SCENE_MODE = 2,
  CAM_STILL_CAPTURE_MODE = 28,
  CAM_FRAME_FORMAT = 30,
  CAM_ALGO_CONTROL = 34,
  CAM_STATUS_CHANGED = 36,
  CAM_BUFFER = 40,
  CAM_RESOLUTION = 31,
  CAM_PIXEL_FORMAT = 30,
  CAM_DIGITAL_ZOOM = 37,
  CAM_VIEWFINDER_MODE = 29,
  CAM_GENERAL_EFFECT_TYPE = 3,
  CAM_LIGHT_FLASH_MODE = 9,
  CAM_EXPOSURE_MANUAL_TIME = 15,
  CAM_EXPOSURE_MANUAL_APERTURE = 18,
  CAM_EXPOSURE_COMPENSATION = 16,
  CAM_EXPOSURE_ISO_VALUE = 17,
  CAM_FOCUS_INFO = 24,
  CAM_OBJECT_TRACING = 32,
} cam_feature_t;

typedef enum {
  CAM_AWB_MODE_AUTO = 0,
  CAM_AWB_MODE_DAYLIGHT = 1,
  CAM_AWB_MODE_CLOUDY = 3,
  CAM_AWB_MODE_SUNSET = 2,
  CAM_AWB_MODE_TUNGSTEN = 4,
  CAM_AWB_MODE_FLUORESCENT = 5,
} cam_awb_mode_t;

typedef enum {
  CAM_GENERAL_EFFECT_TYPE_NORMAL = 0,
  CAM_GENERAL_EFFECT_TYPE_SEPIA = 1,
  CAM_GENERAL_EFFECT_TYPE_NEGATIVE = 2,
  CAM_GENERAL_EFFECT_TYPE_GRAYSCALE = 3,
  CAM_GENERAL_EFFECT_TYPE_NATURAL = 4,
  CAM_GENERAL_EFFECT_TYPE_VIVID = 5,
  CAM_GENERAL_EFFECT_TYPE_COLORSWAP = 6,
  CAM_GENERAL_EFFECT_TYPE_SOLARIZE = 7,
  CAM_GENERAL_EFFECT_TYPE_OUT_OF_FOCUS = 8,
} cam_general_effect_type_t;

typedef enum {
  CAM_GENERAL_SCENE_MODE_AUTO = 4,
  CAM_GENERAL_SCENE_MODE_CLOSEUP = 0,
  CAM_GENERAL_SCENE_MODE_PORTRAIT = 1,
  CAM_GENERAL_SCENE_MODE_LANDSCAPE = 2,
  CAM_GENERAL_SCENE_MODE_SPORT = 3,
  CAM_GENERAL_SCENE_MODE_NIGHT_PORTRAIT = 10,
  CAM_GENERAL_SCENE_MODE_VIDEO_AUTO = 5,
  CAM_GENERAL_SCENE_MODE_VIDEO_NIGHT = 6,
} cam_general_scene_mode_t;

typedef enum {
  CAM_LIGHT_FLASH_MODE_AUTO = 0,
  CAM_LIGHT_FLASH_MODE_OFF = 1,
  CAM_LIGHT_FLASH_MODE_ON = 2,
  CAM_LIGHT_FLASH_MODE_FILL_IN = 3,
  CAM_LIGHT_FLASH_MODE_RED_EYE = 4,
} cam_light_flash_mode_t;

typedef enum {
  CAM_FOCUS_MODE_AUTO = 0,
  CAM_FOCUS_MODE_MACRO = 1,
  CAM_FOCUS_MODE_PORTRAIT = 2,
  CAM_FOCUS_MODE_INFINITY = 3,
  CAM_FOCUS_MODE_HYPERFOCAL = 4,
  CAM_FOCUS_MODE_EXTENDED = 5,
  CAM_FOCUS_MODE_SPORT = 6,
} cam_focus_mode_t;

typedef enum {
  CAM_CONTROL_RUN = 0,
  CAM_CONTROL_STOPED = 1,
} cam_algo_control_t;

typedef enum {
  CAM_STILL_CAPTURE_MODE_HIGH_QUALITY_NO_POSTPROCESS = 0,
  CAM_STILL_CAPTURE_MODE_PERFORMANCE = 4,
} cam_still_capture_mode_t;

typedef enum {
  CAM_LIGHT_PRIVACY_MODE_AUTO = 2,
  CAM_LIGHT_PRIVACY_MODE_ON = 1,
  CAM_LIGHT_PRIVACY_MODE_OFF = 0,
} cam_light_privacy_mode_t;

typedef enum {
  CAM_LIGHT_ASSIST_MODE_AUTO = 0,
  CAM_LIGHT_ASSIST_MODE_ON = 2,
  CAM_LIGHT_ASSIST_MODE_OFF = 1,
} cam_light_assist_mode_t;

typedef enum {
  CAM_EXPOSURE_SHAKE_RISK_LOW = 0,
  CAM_EXPOSURE_SHAKE_RISK_MEDIUM = 1,
  CAM_EXPOSURE_SHAKE_RISK_HIGH = 2,
} cam_exposure_shake_risk_t;

typedef enum { /* 1:1 mapping */
  CAM_FLASH_STATUS_READY = 0,
  CAM_FLASH_STATUS_COOLDOWN = 1,
} cam_light_flash_status_t;

typedef enum {
  CAM_FOCUS_STATUS_IDLE = 0,
  CAM_FOCUS_STATUS_RUNNING = 1,
  CAM_FOCUS_STATUS_SUCCESS = 2,
  CAM_FOCUS_STATUS_FAIL = 3,
} cam_focus_status_t;

typedef enum {
  CAM_GENERAL_FLICKER_REDUCTION_MODE_OFF = 0,
  CAM_GENERAL_FLICKER_REDUCTION_MODE_50HZ = 1,
  CAM_GENERAL_FLICKER_REDUCTION_MODE_60HZ = 2,
  CAM_GENERAL_FLICKER_REDUCTION_MODE_AUTO = 3,
} cam_general_flicker_reduction_mode_t;

typedef enum {
  CAM_OBJECT_ORIENTATION_UNKNOWN = 0,
  CAM_OBJECT_ORIENTATION_0 = 1,
  CAM_OBJECT_ORIENTATION_90 = 2,
  CAM_OBJECT_ORIENTATION_270 = 3,
} cam_object_orientation_t;

typedef enum {
  CAM_IPP_VERSION_1_8 = 0,
  CAM_IPP_VERSION_1_9 = 1,
} cam_ipp_version_t;

typedef enum {
  UNKNOWN = 0,
  AVERAGE = 1,
  CENTER_WEIGHTED_AVERAGE = 2,
  SPOT = 3,
  MULTI_SPOT = 4,
  PATTERN = 5,
  PARTIAL = 6,
} cam_metering_mode_t;

typedef enum {
  NOT_EXIST = 0,
  Y_COMPONENT = 1,
  CB_COMPONENT = 2,
  CR_COMPONENT = 3,
  R_COMPONENT = 4,
  G_COMPONENT = 5,
  B_COMPONENT = 6,
} cam_components_configuration_t;

typedef enum {
  AUTO_EXPOSURE = 0,
  MANUAL_EXPOSURE = 1,
  AUTO_BRACKET = 2,
} cam_exposure_mode_t;

typedef enum {
  AUTO_WHITE_BALANCE = 0,
  MANUAL_WHITE_BALANCE = 1,
} cam_white_balance_t;

typedef enum {
  STANDART = 0,
  LANDSCAPE = 1,
  PORTRAIT = 2,
  NIGHT_SCENE = 3,
  MACRO = 4,
  ACTION = 5,
} cam_scene_capture_type_t;

typedef enum {
  CAM_LIGHT_FLASH_IS_REQUIRED, // TODO: value?
} cam_light_flash_required_t;

void *cam_library_create();
void cam_library_delete(void *handle);
cam_err_t cam_feature_set(void *handle, cam_feature_t feature, ...);
cam_err_t cam_feature_get(void *handle, cam_feature_t feature, ...);

typedef struct {
  void (*callback)(void *);
  void *argument;
} cam_status_changed_t;

typedef struct {
  unsigned int coverage;
  unsigned int zones_vertical;
  unsigned int zones_horizontal;
  unsigned int zone;
  unsigned int lens_position;
} cam_focus_info_t;

typedef struct {
  unsigned int numenator;
  unsigned int denumenator;
} cam_fractional_t;

typedef struct {
  cam_fractional_t exposure_time;
  cam_fractional_t fnumber;
  unsigned int iso_speed_ratings;
  cam_components_configuration_t components_configuration[4];
  cam_fractional_t shutter_speed_value;
  cam_fractional_t aperture_value;
  cam_metering_mode_t metering_mode;
  unsigned int light_source;
  unsigned int flash;
  cam_fractional_t focal_length;
  unsigned int pixel_x_dimension;
  unsigned int pixel_linesize;
  unsigned int pixel_y_dimension;
  cam_exposure_mode_t exposure_mode;
  cam_white_balance_t white_balance;
  cam_fractional_t digital_zoom_ratio;
  cam_scene_capture_type_t scene_capture_type;
  unsigned int gain;
  int contrast;
  int saturation;
  int sharpness;
  unsigned int sensor_gain_ev;
  unsigned int lens_position;
  unsigned int reserved[4];
} frame_info_t;

typedef struct {
  void *b_address;
  unsigned int b_size;
  cam_err_t status;
  cam_status_changed_t notify;
  unsigned int index;
} cam_frame_buffer_t;

typedef struct {
  int top;
  int left;
  int width;
  int height;
  int bytesperline;
  int size;
} cam_window_rect_t;

typedef struct {
  float x_vector;
  float y_vector;
  int d_time;
} cam_motion_vector_t;

typedef struct {
  cam_window_rect_t frame_size;
  cam_window_rect_t objects[OBJECT_TRACKING_MAX_COUNT];
  int priority[OBJECT_TRACKING_MAX_COUNT];
  cam_object_orientation_t rotation;
} cam_object_tracking_t;

// break gst_subdevsrc_libomap3camd_get_makernote
// break g_malloc
// info registers -> r0
unsigned char maker_note_t[4096];

typedef struct {
  unsigned int low;
  unsigned int mid;
  unsigned int high;
} cam_filter_strength_t;

typedef struct {
  unsigned int luma_noise_filter;
  unsigned int chroma_noise_filter;
  unsigned int edge_enhancement_strength;
  unsigned int weak_edge_threshold;
  unsigned int strong_edge_threshold;
} cam_ipp_paramters_1_8;

typedef struct {
  unsigned int edge_enhancement_strength;
  unsigned int weak_edge_threshold;
  unsigned int strong_edge_threshold;
  cam_filter_strength_t freq_luma_noise;
  cam_filter_strength_t freq_cb_noise;
  cam_filter_strength_t freq_cr_noise;
  unsigned int shading_vert_param1;
  unsigned int shading_vert_param2;
  unsigned int shading_horz_param1;
  unsigned int shading_horz_param2;
  unsigned int shading_gain_scale;
  unsigned int shading_gain_offset;
  unsigned int shading_gain_maxvalue;
  unsigned int ratio_downsample_cb_cr;
} cam_ipp_paramters_1_9;

typedef struct {
  cam_ipp_version_t version;
  union {
    cam_ipp_paramters_1_8 ipp_1_8;
    cam_ipp_paramters_1_9 ipp_1_9;
  } info;
} cam_ipp_parameters_t;

#endif /* __OMAP3_CAM_H__ */
